intermediate_notebooks/E2E/taxi/NYCTaxi-E2E-GPU-Dask.ipynb

```diff
@@ -0,0 +1,1904 @@
+    "                ddf[col] = ddf[col].astype('float32')\n",
+    "            ddf[col] = ddf[col].fillna(-1)\n",
+    "    \n",
+    "    return ddf"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<b> NOTE: </b>We will realize that some of 2015 data has column name as `RateCodeID` and others have `RatecodeID`. When we rename the columns in the clean function, it internally doesn't pass meta while calling map_partitions(). This leads to the error of column name mismatch in the returned data. For this reason, we will call the clean function with map_partition and pass the meta to it. Here is the link to the bug created for that: https://github.com/rapidsai/cudf/issues/5413 "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "df_2014 = df_2014.map_partitions(clean, must_haves, meta=must_haves)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We still have 2015 and the first half of 2016's data to read and clean. Let's increase our dataset."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "df_2015 = dask_cudf.read_csv(base_path+'2015/yellow_*.csv')"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "df_2015 = df_2015.map_partitions(clean, must_haves, meta=must_haves)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Handling 2016's Mid-Year Schema Change"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In 2016, only January - June CSVs have the columns we need. If we try to read base_path+2016/yellow_*.csv, Dask will not appreciate having differing schemas in the same DataFrame.\n",
+    "\n",
+    "Instead, we'll need to create a list of the valid months and read them independently."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "months = [str(x).rjust(2, '0') for x in range(1, 7)]\n",
+    "valid_files = [base_path+'2016/yellow_tripdata_2016-'+month+'.csv' for month in months]"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#read & clean 2016 data and concat all DFs\n",
+    "df_2016 = dask_cudf.read_csv(valid_files).map_partitions(clean, must_haves, meta=must_haves)\n",
+    "\n",
+    "#concatenate multiple DataFrames into one bigger one\n",
+    "taxi_df = dask.dataframe.multi.concat([df_2014, df_2015, df_2016])"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "taxi_df = taxi_df.persist()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Exploratory Data Analysis (EDA)"
```

Comment:
@think-high I'm eager to see what you can do here using ops that run on GPU, instead of falling back to Pandas.